server/client.go

```diff
@@ -1966,6 +2001,62 @@ func (c *client) canSubscribe(subject string) bool {
 	return allowed
 }
 
+func queueMatches(queue string, qsubs [][]*subscription) bool {
+	if len(qsubs) == 0 {
+		return true
+	}
+	for _, qsub := range qsubs {
+		qs := qsub[0]
+		qname := string(qs.queue)
+
+		// NOTE: '*' and '>' tokens can also be valid
+		// queue names so we first check against the
+		// literal name.  e.g. v1.* == v1.*
+		if queue == qname || (subjectHasWildcard(qname) && subjectIsSubsetMatch(queue, qname)) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *client) canQueueSubscribe(subject, queue string) bool {
+	if c.perms == nil {
+		return true
+	}
+
+	allowed := true
+	if c.perms.sub.allow != nil {
+		r := c.perms.sub.allow.Match(subject)
+
+		// If there is an explicit sub permission it allows any
+		// queue subscription unless there are explicit rules
+		// involving queue subscriptions.
+		subAllowed := len(r.psubs) != 0
```

Comment:
Sorry to get back to that, but I had to re-read the condition below to understand what was happening.
Is that correct to say that since we are here checking a subscribe permission for a queue, that it is allowed only if there is no queue specific permissions, or if there is, the queue has to be in the list?
If so, I wonder if all that could not be simplified to:
```
if len(r.qsubs) > 0 {
    allowed = queueMatches(queue, r.qsubs)
} else {
   allowed = len(r.psubs) != 0
}
```
The first `if` means that we found explicit queue permissions, so we make sure that the queue name `queue` is in that list. If there is no specific queue permissions, we fallback to the original `allowed = len(r.psubs) != 0`.

Reply:
Yep, good catch!